### ISSUE TITLE
Android building with NDK-21d fails

### ISSUE THREAD

**Issue by mgorges** [part-cw/lambdanative#309]
@karliwalti  wrote: "Ok, updated to ndk-21d, must say this build process starts to explode my harddisk. 60GB is not enough  to host an ubuntu and android build pipepline... nevertheless building DemoHelloWorld for Android 21, 24 and 29 fails with similar error messages:

```make
=== using profile karliwalti [/home/walterk/SoftwareRepo/lambdanative/PROFILE]..
=== configured to build DemoHelloWorld version 1.0 for android on linux in normal mode

ls: cannot access '/home/walterk/.cache/lambdanative/android/android-ndk-r21d-arm64-21-toolchain/bin/*-gcc': No such file or directory
==> checking for required tools..
==> checking for required libraries..
==> checking for lambdanative tools..
ls: cannot access '/home/walterk/.cache/lambdanative/android/android-ndk-r21d-arm-21-toolchain/bin/*-gcc': No such file or directory
==> creating libraries needed for DemoHelloWorld..
 => liblambdanative..
 => cleaning up..
 => exploding library liblambdanative..
==> creating artwork needed for DemoHelloWorld..
==> creating textures needed for DemoHelloWorld..
==> creating fonts needed for DemoHelloWorld..
==> creating strings needed for DemoHelloWorld..
==> updating embedded files for DemoHelloWorld..
==> updating web assets for DemoHelloWorld..
 => compiling scheme payload..
 => done compiling scheme payload
 => generating hook..
 => assembling payload..
 == /home/walterk/.cache/lambdanative/android/arm/lib/libpayload.a
ls: cannot access '/home/walterk/.cache/lambdanative/android/android-ndk-r21d-arm64-21-toolchain/bin/*-gcc': No such file or directory
==> creating libraries needed for DemoHelloWorld..
 => liblambdanative..
 => cleaning up..
 => exploding library liblambdanative..
==> creating artwork needed for DemoHelloWorld..
==> creating textures needed for DemoHelloWorld..
==> creating fonts needed for DemoHelloWorld..
==> creating strings needed for DemoHelloWorld..
==> updating embedded files for DemoHelloWorld..
==> updating web assets for DemoHelloWorld..
 => compiling scheme payload..
 => done compiling scheme payload
 => generating hook..
 => assembling payload..
 == /home/walterk/.cache/lambdanative/android/arm64/lib/libpayload.a
==> creating android loader needed for DemoHelloWorld..
 => creating android project..
 => using target 1 [API 21]
 => preparing icons..
 => processing sounds needed for DemoHelloWorld..
 => processing sounds from /home/walterk/SoftwareRepo/lambdanative/apps/DemoHelloWorld/sounds...
 => /home/walterk/SoftwareRepo/lambdanative/apps/DemoHelloWorld/sounds/ahooga.wav..
 => transferring java hook..
 => transferring java public classes..
 => preparing manifest..
 => creating payload module..
 => preparing JNI..
 => compiling payload module..
NDK_MODULE_PATH=/home/walterk/.lambdanative/tmp.qoolES /usr/local/android-ndk-r21d/ndk-build --silent APP_ABI=" armeabi-v7a arm64-v8a"
Android NDK: Found platform level in ./project.properties. Setting APP_PLATFORM to android-21.    
[armeabi-v7a] Compile thumb  : payloadshared <= bootstrap.c
jni/bootstrap.c:143:12: warning: returning 'const char *' from a function with result type 'char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
    return str;
           ^~~
1 warning generated.
[armeabi-v7a] SharedLibrary  : libpayloadshared.so
/usr/local/android-ndk-r21d/toolchains/llvm/prebuilt/linux-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: /home/walterk/.lambdanative/tmp.qoolES/libpayload/armeabi-v7a/libpayload.a(1959055257.o): incompatible target
/usr/local/android-ndk-r21d/toolchains/llvm/prebuilt/linux-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: /home/walterk/.lambdanative/tmp.qoolES/libpayload/armeabi-v7a/libpayload.a(2907813665.o): incompatible target
jni/bootstrap.c:54: error: undefined reference to 'ffi_event'
jni/bootstrap.c:59: error: undefined reference to 'ffi_event'
jni/bootstrap.c:119: error: undefined reference to 'scm_mediascanner_callback'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [/usr/local/android-ndk-r21d/build/core/build-binary.mk:725: obj/local/armeabi-v7a/libpayloadshared.so] Error 1
ERROR: failed with exit code 2
BUILD FAILED - configure with verbose option for more information
make: *** [Makefile:2: all] Error 1
```

**Comment by mgorges** [comment]
I have moved this to a different issue as this has nothing to do with the ANDROIDSDK being open? Based on this last piece, you cannot use **ndk-21d**, as it doesn't include GCC so tests fail. Also the `incompatible target` is likely the culprit of the problem, as the path includes `armeabi-v7a` yet the toolchain states `arm64`.

**Comment by mgorges** [comment]
The second time you see ```[armeabi-v7a] SharedLibrary  : libpayloadshared.so``` it should read ```[arm64-v8a] Compile        : payloadshared <= bootstrap.c```

**Comment by peterlew** [comment]
@karliwalti I get a very similar error when switching between APIs for builds. Usually if I am building for a higher API and then switch to a lower one. Deleting the "android" folder in Caches/lambdanative always works for me, maybe it would work for you too? I'll try to find a more satisfying fix as well.

**Comment by karliwalti** [comment]
true. I repeated again the full installation on a clean system and it compiles differently.  However, I still get errors with respect to the  keystore. Is the ndk version related in any way to the signing of the app?

**Comment by karliwalti** [comment]
this is now also solved, see #314

**Comment by mgorges** [comment]
Closing as this specific issue is resolved; a general purpose one might be needed, see #315
